### PR TITLE
[ENH] Add case_when #736

### LINKED ISSUE #736: Case_when function
# Brief Description

I would like to propose a `case_when` function, similar to the same function in SQL and R, for conditions. It will be a wrapper around `np.select`.

# Example API
```python
       Type	Set
1	A	Z
2	B	Z
3	B	X
4	C	Y
```

```python
# run checks based on conditions and create a new column, or modify existing column.
@pf.register_dataframe_method
def case_when(df, conditions,column_name, default = None, ) : 
    condlist, choicelist = zip(*[(value,key) for key, value in conditions.items()])
    df = df.assign(**{column_name : np.select(condlist, choicelist, default=default)})
    return df

cond = {"green" : df.Set=="Z", "yellow":df.Set=="X"}
df.case_when(conditions=cond, default=None, column_name='rag')

       Type	Set	rag
1	A	Z	green
2	B	Z	green
3	B	X	yellow
4	C	Y	None
```

### PR DESCRIPTION
**Pull request by robertmitchellv**
# Adding a `case_when` function

Please describe the changes proposed in the pull request:

- Effectively, this is a wrapper around `pd.Series.mask` but with an easier to manage mechanism for passing arguments using variable arguments.
- chose `pd.Series.mask` to allow support for pandas extension arrays/dtypes (np.select would convert them to numpy dtypes, which wont be beneficial for nullable Integer dtypes, or String dtypes, or other future pandas dtypes).
- The function does not mutate the original `pd.DataFrame` and is vectorized.
- Can be method chained

**This PR resolves #736 **

Speed comparison # ... with a pinch of salt ... 
```py
df = pd.DataFrame({'col1': list('ABBC'), 'col2': list('ZZXY')})
  col1 col2
0    A    Z
1    B    Z
2    B    X
3    C    Y

df = pd.concat([df]*100_000)

df.shape
(400000, 2)


%%timeit
conditions = [
    (df['col2'] == 'Z') & (df['col1'] == 'A'),
    (df['col2'] == 'Z') & (df['col1'] == 'B'),
    (df['col1'] == 'B')
]

choices = ['yellow', 'blue', 'purple']

df['color'] = np.select(conditions, choices, default='black')

167 ms ± 2.91 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


df.case_when(
    df.col2.eq("Z") & df.col1.eq("A"), "yellow",  # first condition and value
    df.col2.eq("Z") & df.col1.eq("B"), "blue",    # second condition and value
    df.col1.eq("B"), "purple",                    # third condition and value 
    "black",                                      # default if no condition is True
    column_name = "color"
    )

125 ms ± 449 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


# anonymous function
df.case_when(
    lambda df: df.col2.eq("Z") & df.col1.eq("A"), "yellow", 
    lambda df: df.col2.eq("Z") & df.col1.eq("B"), "blue",    
    lambda df: df.col1.eq("B"), "purple",                         
    "black",                                                     
    column_name = "color"
    )

134 ms ± 2.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


# pd.eval
%%timeit
df.case_when(
   "col2 == 'Z' and col1 == 'A'", "yellow", 
  "col2 == 'Z' and col1 == 'B'", "blue",   
 "col1 == 'B'", "purple",                        
   "black",                                             
  column_name = "color"
  )
74.4 ms ± 4.46 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl
- @samukweku

### CODE REVIEW

**Comment by robertmitchellv** [comment]
I'm not exactly sure how to _add tests_ here--is it the `coverage.xml` file? I ran `pytest janitor/functions.py` and everything seemed to check out.

**Comment by samukweku** [comment]
@robertmitchellv There is a tests folder. Look for the functions subfolder and add a test file there. You could have a look at existing files as a guide.

**Comment by samukweku** [comment]
@robertmitchellv , this may be a useful [link](https://towardsdatascience.com/efficient-implementation-of-conditional-logic-on-pandas-dataframes-4afa61eb7fce)

**Comment by ericmjl** [comment]
@robertmitchellv thank you for the PR! My apologies for responding back late. Our first child just arrived, and I've been head underwater catching up with everything in life.

Because @samukweku has thought more about `case_when` than I have, I'm going to defer to his judgment here on the exact details of the implementation.

If possible, would you be kind enough to add in some tests, please? The best place for the tests is in the `tests/functions/` directory, inside a `test_case_when.py` file. It's something we ask for, to maintain high test coverage of the library (this gives our end-users confidence that the functions behave as expected).

A final small friendly note -- you'll see the docs preview show up, but they build on every push to the PR, which does eat up into the build minutes on Netlify. I've been paying out of pocket for the build minutes (about $7/mo once I run out of the free tier that's associated with my account). So if possible, I'll be super appreciative of it if you could push "conservatively", i.e. push up a few commits at the same time, rather than push every commit. (And in case I come off the wrong way, this isn't a comment targeted towards you at all! I really appreciate the time you've spent crafting the PR. :smile: Just wanted to make sure you're aware of some of the constraints we're operating with in the project.)

**Comment by samukweku** [comment]
@ericmjl , I thought you had cancelled the netliffy charge? personal opinion though, I think you should. We can go back to running `make docs`, view it on our local machine, before pushing to github.

**Comment by ericmjl** [comment]
I was thinking seriously about cancelling, but I later realized, actually it's not too bad. And I do use Netlify for other sites myself, like my personal site. It's two coffees worth per month, which I am happy to pay for the convenience 😃

**Comment by ericmjl** [comment]
@robertmitchellv there's some good stuff in here! Are you still planning to work on the PR?

**Comment by robertmitchellv** [comment]
@ericmjl sorry for the late response! I am going to finish this, but got swamped at work. I'll try to set time aside this week to push it through.

Also congrats on your first child 🎉🍾 so exciting

**Comment by robertmitchellv** [comment]
@samukweku Sorry for how long it took to add this! Work has been all consuming! Let me know if the test I added isn't sufficient. It passed, but I'm sure it could be better.

File: janitor/functions.py
Comment (samukweku) [inline janitor/functions.py]: I think we should make the syntax look a bit more like numpy select than dplyr's case_when; more familiar to python users; the True makes sense within R, not really within python. I'll suggest using `default`.

File: janitor/functions.py
Comment (samukweku) [inline janitor/functions.py]: I think the part we also need to test is how it works for a workflow, where we have maybe four or more operations before the case_when is called. How will `case_when` be able to refer to the current dataframe? will `df` have changed? will the `df` being referred to here mean the original `df`? the current `df`? Also think it would help if we could pass a string or a callable, maybe it would help with aesthetics? maybe you could add some tests for scenarios where the workflow did not start with `case_when`, to see how it goes.

File: tests/functions/test_case_when.py
Comment (samukweku) [inline tests/functions/test_case_when.py]: I dont think the description here is right, as this is not in the query-style form. This is more pure python. query-style would be more like the example in [janitor.update_where](https://pyjanitor.readthedocs.io/reference/janitor.functions/janitor.update_where.html#janitor.update_where) or pandas' [query](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html) function.

File: tests/functions/test_case_when.py
Comment (samukweku) [inline tests/functions/test_case_when.py]: something else that comes to mind, is testing if numpy select preserves the dtype of the column - specifically pandas extension arrays. Say the column is a category dtype, will numpy select preserve it? I dont think so. what if it is an integer column with pandas nullable integer. will it be preserved? I dont think so. I think it will revert to numpy's own dtype. Thinking about this makes me wonder if np.select is really the best tool for the job. I know an [issue](https://github.com/pandas-dev/pandas/issues/39154) has been raised in Pandas regarding `case_when`; however, it is still an issue, with no real traction.

**Review by samukweku — CHANGES_REQUESTED** [review]
@robertmitchellv     glad you could work on it. made some comments. hope they are constructive.

**Comment by ericmjl** [comment]
> @samukweku Sorry for how long it took to add this! Work has been all consuming! Let me know if the test I added isn't sufficient. It passed, but I'm sure it could be better.

No worries, @robertmitchellv! Tis all good, that's just the nature of OSS :smile:. Glad you could come back to the PR!

Apart from @samukweku's recent comments, could you also merge in the latest changes in `dev` into your branch? That should resolve most of the conflicts present.

**Comment by samukweku** [comment]
@robertmitchellv @ericmjl I'd love to take a stab at your PR and make some changes (got a few ideas I would love to tinker and push through your PR) and get it over the finish line

**Comment by robertmitchellv** [comment]
@samukweku that would be wonderful! I was struggling to figure out how to incorporate your suggestions and then got buried with things at work. I would find a lot of value in seeing how you'd solve for the issues that remain here--I know @ericmjl asked me to merge latest changes--I'll have to look at what I have locally here because I don't know if I've made many changes outside of what I've already pushed (I may also not understand the ask and that could be where the confusion stems).

**Comment by samukweku** [comment]
@robertmitchellv no worries. Started a new job as well ... been a busy one too - busy is good. I cant push commits to your branch, It seems you unchecked the box that says `Allow edits from maintainers`. Could you kindly check that box? Thanks

**Comment by robertmitchellv** [comment]
@samukweku sorry about that! Didn't realize I had that unchecked 😅 and congrats on the new job! That's great! Hopefully you won't run into any other issues--I'll try to get them resolved as soon as I can though

**Comment by samukweku** [comment]
thanks @robertmitchellv ; for some reason, I still cant make any commits. @ericmjl any ideas on how I can pull robert's PR to my local branch, make commits, and push back into his PR? I guess you'll have some git-fu magic

**Comment by samukweku** [comment]
@robertmitchellv ; picking this up to continue; let me know if you are back on it and I'll defer to you.

**Comment by samukweku** [comment]
@robertmitchellv, got some updates on the `case_when` function that I'd like you to have a look at; I couldnt update your PR. @ericmjl I might need your git skills again to get into @robertmitchellv PR

**Comment by samukweku** [comment]
@robertmitchellv made some edits to the code, and added some tests. Please feel free to make changes.

**Comment by robertmitchellv** [comment]
@samukweku this looks great! Sorry I wasn't able to push it over the finish line. Very grateful to you!

File: janitor/functions.py
Comment (ericmjl) [inline janitor/functions.py:6901]: Good choice here to use sphinx-style blocks for now.

**Review by ericmjl — APPROVED** [review]
:100: from me, @samukweku! Thank you for handling this PR. Also, thank you @robertmitchellv for initiating the idea!

**Comment by ericmjl** [comment]
@samukweku I'm going to merge. Thank you sooooooooo much for taking this over the finish line. :tada:

@robertmitchellv thank you too for initiating the PR! :100:% great feature to have in the library.

**Comment by robertmitchellv** [comment]
@ericmjl thank you! Big thanks to @samukweku 🙏🏻 

This might be my first real OSS contribution 😅

Appreciate how welcoming and understanding you've both been; great community here 👍🏻

**Pull request merged**
